src/Model/Item.php

```diff
@@ -422,4 +1067,659 @@ public static function updateContact($arr) {
+			dbesc($c[0]['url']),
+			dbesc($c[0]['thumb']),
+			intval($private),
+			dbesc($u[0]['allow_cid']),
+			dbesc($u[0]['allow_gid']),
+			dbesc($u[0]['deny_cid']),
+			dbesc($u[0]['deny_gid']),
+			intval($item_id)
+		);
+		update_thread($item_id);
+
+		Worker::add(['priority' => PRIORITY_HIGH, 'dont_fork' => true], 'Notifier', 'tgroup', $item_id);
+
+	}
+
+	public static function isRemoteSelf($contact, &$datarray) {
+		$a = get_app();
+
+		if (!$contact['remote_self']) {
+			return false;
+		}
+
+		// Prevent the forwarding of posts that are forwarded
+		if ($datarray["extid"] == NETWORK_DFRN) {
+			return false;
+		}
+
+		// Prevent to forward already forwarded posts
+		if ($datarray["app"] == $a->get_hostname()) {
+			return false;
+		}
+
+		// Only forward posts
+		if ($datarray["verb"] != ACTIVITY_POST) {
+			return false;
+		}
+
+		if (($contact['network'] != NETWORK_FEED) && $datarray['private']) {
+			return false;
+		}
+
+		$datarray2 = $datarray;
+		logger('remote-self start - Contact '.$contact['url'].' - '.$contact['remote_self'].' Item '.print_r($datarray, true), LOGGER_DEBUG);
+		if ($contact['remote_self'] == 2) {
+			$r = q("SELECT `id`,`url`,`name`,`thumb` FROM `contact` WHERE `uid` = %d AND `self`",
+				intval($contact['uid']));
+			if (DBM::is_result($r)) {
+				$datarray['contact-id'] = $r[0]["id"];
+
+				$datarray['owner-name'] = $r[0]["name"];
+				$datarray['owner-link'] = $r[0]["url"];
+				$datarray['owner-avatar'] = $r[0]["thumb"];
+
+				$datarray['author-name']   = $datarray['owner-name'];
+				$datarray['author-link']   = $datarray['owner-link'];
+				$datarray['author-avatar'] = $datarray['owner-avatar'];
+
+				unset($datarray['created']);
+				unset($datarray['edited']);
+			}
+
+			if ($contact['network'] != NETWORK_FEED) {
+				$datarray["guid"] = get_guid(32);
+				unset($datarray["plink"]);
+				$datarray["uri"] = item_new_uri($a->get_hostname(), $contact['uid'], $datarray["guid"]);
+				$datarray["parent-uri"] = $datarray["uri"];
+				$datarray["extid"] = $contact['network'];
+				$urlpart = parse_url($datarray2['author-link']);
+				$datarray["app"] = $urlpart["host"];
+			} else {
+				$datarray['private'] = 0;
+			}
+		}
+
+		if ($contact['network'] != NETWORK_FEED) {
+			// Store the original post
+			$r = self::insert($datarray2, false, false);
+			logger('remote-self post original item - Contact '.$contact['url'].' return '.$r.' Item '.print_r($datarray2, true), LOGGER_DEBUG);
+		} else {
+			$datarray["app"] = "Feed";
+		}
+
+		// Trigger automatic reactions for addons
+		$datarray['api_source'] = true;
+
+		// We have to tell the hooks who we are - this really should be improved
+		$_SESSION["authenticated"] = true;
+		$_SESSION["uid"] = $contact['uid'];
+
+		return true;
+	}
+
+	/**
+	 *
+	 * @param string $s
+	 * @param int    $uid
+	 * @param array  $item
+	 * @param int    $cid
+	 * @return string
+	 */
+	public static function fixPrivatePhotos($s, $uid, $item = null, $cid = 0)
```

Comment:
Please consider moving this to `BBCode::fixPrivatePhotos`

Reply:
I had been thinking about this - but it depends on the functions below that I had to move as well - but they do more belong to items.

Reply:
I saw that, but the fact that tipped me is that the main parameter of this function is a BBCode string.

Reply:
But I wouldn't like to move the other functions to BBCode - but I want to keep the structures always as local as possible.

Reply:
I understand, that's why it wasn't a strong imperative.